Source/CCZ4/Weyl4.impl.hpp

```diff
@@ -123,23 +138,40 @@ Weyl4::compute_EB_fields(const Vars<data_t> &vars,
     // Calculate electric and magnetic fields
     FOR2(i, j)
     {
-        out.E[i][j] = 0;
-        out.B[i][j] = 0;
+        out.E[i][j] = 0.0;
+        out.B[i][j] = 0.0;
     }
 
     FOR4(i, j, k, l)
     {
         out.B[i][j] +=
-            epsilon3_LUU[i][k][l] * (covariant_deriv_K_tensor[l][j][k]);
+            epsilon3_LUU[i][k][l] * covariant_deriv_K_tensor[l][j][k];
     }
 
-    FOR2(i, j) { out.E[i][j] += ricci.LL[i][j] + vars.K * K_tensor[i][j]; }
+    FOR2(i, j)
+    {
+        out.E[i][j] += ricci_and_Z_terms.LL[i][j] + vars.K * K_tensor[i][j];
+    }
 
     FOR4(i, j, k, l)
     {
         out.E[i][j] += -K_tensor[i][k] * K_tensor[l][j] * h_UU[k][l] * vars.chi;
     }
 
+    if (m_formulation == CCZ4RHS<>::USE_CCZ4)
+    {
+        // In CCZ4 case do explicit symmetrization; BSSN case relies on momentum
+        // constraint satisfaction instead
+        TensorAlgebra::make_symmetric(out.B);
+
+        FOR2(i, j) { out.E[i][j] += -vars.Theta * K_tensor[i][j]; }
+
+        // The expression in CCZ4 is explicitly trace-free but in BSSN it's only
+        // trace-free if the Hamiltonian constraint is satisfied...
+        // Let's keep this CCZ4 only to avoid breaking the test
+        TensorAlgebra::make_trace_free(out.E, vars.h, h_UU);
+    }
+
```

Comment:
This is the big discussion. We'll probably need external input (@KAClough ?).
In BSSN, should we force 1) Eij to be traceless and 2) Bij to be symmetric or not.
Against:
- not doing it is leaving them as they actually come out when deriving them
- this also means leaving them as Albubierre has them in the book
- not doing it means keeping the old behaviour for BSSN
- maybe some more, but I am clearly biased, apologies

In favour:
- Eij and Bij should be geometrically naturally explicitly symmetric and traceless. They are not because we use Einstein equations in the derivation and that breaks those properties. They are still true, but only reestablished once we use the constraint equations as @mirenradia pointed out. Forcing their symmetry is equivalent to adding a multiple of the constraints, which should be valid in BSSN and give us those nice properties. As the properties are not explicit, this means constraint violations mean that Eij will not be traceless and Bij will not be symmetric.
- I started doing these CCZ4 versions of Eij and Bij because I was suffering from this problem. If one does further calculations with these tensors (e.g. compute the Kretschmann) then the result will depend (in terms of numerical precision) on the order in which the indices are put (`Bij[i][j]` vs `Bij[j][i]`). That's bad.
- I've had to store Eij and Bij. But we only store the top diagonal. So if Bij is not symmetric, when we read it back again (from hdf5's) we will read it as symmetric. This means the old code state of the tensor is 1) not recoverable 2) the recovered version will be symmetric but probably with larger errors than it already had with the properties violated. 
- It avoids having to add matter terms to Bij in `MatterWeyl`, which are 0 if Bij is explicitly symmetrized. (don't forget to change this if this goes forward)
- As your 3.16 formula in the pdf shows, the Eij would also be missing a cosmological constant term. Which doesn't make any sense because 1) we wouldn't add it because our GW extraction is for asymptotically flat and 2) it just adds to the non-tracelessness that Eij should NOT have. So this just makes me more uncomfortable with the inconsistency

I feel only averagely strongly about this in the sense that I never use BSSN. So it's about the code. From the problems that come from it I think it's definitely worth it and not such a dramatic change to enforce symmetry of Bij and tracelessness of Eij, which is nothing more than adding a multiple of the constraints.

Reply:
>I started doing these CCZ4 versions of Eij and Bij because I was suffering from this problem. If one does further calculations with these tensors (e.g. compute the Kretschmann) then the result will depend (in terms of numerical precision) on the order in which the indices are put (Bij[i][j] vs Bij[j][i]). That's bad.

I see your point but how does one know which is closer to the true value anyway (i.e. before or after symmetrization)?

>I've had to store Eij and Bij. But we only store the top diagonal. So if Bij is not symmetric, when we read it back again (from hdf5's) we will read it as symmetric. This means the old code state of the tensor is 1) not recoverable 2) the recovered version will be symmetric but probably with larger errors than it already had with the properties violated.

This would be annoying but is probably not applicable to non-modified gravity?

>As your 3.16 formula in the pdf shows, the Eij would also be missing a cosmological constant term. Which doesn't make any sense because 1) we wouldn't add it because our GW extraction is for asymptotically flat and 2) it just adds to the non-tracelessness that Eij should NOT have. So this just makes me more uncomfortable with the inconsistency.

I don't think this is a problem as no one should use these classes for non-asymptotically flat spacetimes anyway.

Another position against is that it would break the test.